server/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1741,4 +1605,98 @@ public final ResponseHandlers getResponseHandlers() {
     public final RequestHandlerRegistry<? extends TransportRequest> getRequestHandler(String action) {
         return requestHandlers.get(action);
     }
+
+    private final class ChannelsConnectedListener implements ActionListener<Void> {
+
+        private static final int FAILED = -1;
+
+        private final DiscoveryNode node;
+        private final ConnectionProfile connectionProfile;
+        private final List<TcpChannel> channels;
+        private final ActionListener<NodeChannels> listener;
+        private final AtomicInteger pendingConnections;
+
+        private ChannelsConnectedListener(DiscoveryNode node, ConnectionProfile connectionProfile, List<TcpChannel> channels,
+                                          ActionListener<NodeChannels> listener) {
+            this.node = node;
+            this.connectionProfile = connectionProfile;
+            this.channels = channels;
+            this.listener = listener;
+            this.pendingConnections = new AtomicInteger(channels.size());
+        }
+
+        @Override
+        public void onResponse(Void v) {
+            assert pendingConnections.get() != 0 : "Should not call onResponse when the pending connections is 0.";
+            // Returns true if all connections have completed successfully
+            if (setConnectSuccess()) {
+                final TcpChannel handshakeChannel = channels.get(0);
+                try {
+                    executeHandshake(node, handshakeChannel, connectionProfile.getHandshakeTimeout(), new ActionListener<Version>() {
+                        @Override
+                        public void onResponse(Version version) {
+                            NodeChannels nodeChannels = new NodeChannels(node, channels, connectionProfile, version);
+                            nodeChannels.channels.forEach(ch -> ch.addCloseListener(ActionListener.wrap(nodeChannels::close)));
+                            listener.onResponse(nodeChannels);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            CloseableChannel.closeChannels(channels, false);
```

Comment:
`CloseableChannel.closeChannel` can throw so should we call the listener first or use try/finally?

Reply:
`CloseableChannel.closeChannel` actually should not throw. `CloseableChannel` `close` signature does not include an exception.

```
    static <C extends CloseableChannel> void closeChannels(List<C> channels, boolean blocking) {
        try {
            IOUtils.close(channels);
        } catch (IOException e) {
            // The CloseableChannel#close method does not throw IOException, so this should not occur.
            throw new UncheckedIOException(e);
        }
        if (blocking) {
            ArrayList<ActionFuture<Void>> futures = new ArrayList<>(channels.size());
            for (final C channel : channels) {
                PlainActionFuture<Void> closeFuture = PlainActionFuture.newFuture();
                channel.addCloseListener(closeFuture);
                futures.add(closeFuture);
            }
            blockOnFutures(futures);
        }
    }
```

Should I changed that to `UncheckedIOException` to `AssertionError`? I can still do `try/catch` if you like for safety.